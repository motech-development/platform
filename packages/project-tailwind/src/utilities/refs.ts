import { LegacyRef, MutableRefObject, useMemo, useRef } from 'react';

/**
 * Check if ref is a mutable ref object
 *
 * @param ref - Reference object or function
 *
 * @returns If ref is a mutable ref object
 */
function isMutableRefObject<T>(
  ref: MutableRefObject<T> | LegacyRef<T>,
): ref is MutableRefObject<T | null> {
  return ref != null;
}

/** useMergeRefs arguments */
type TUseMergeRefsInput<T> = Array<MutableRefObject<T> | LegacyRef<T>>;

/**
 * Combines multiple refs into a single memoized callback instance
 *
 * @param refs - Reference object or function
 *
 * @returns Combined ref object
 */
export default function useMergeRefs<T>(...refs: TUseMergeRefsInput<T>) {
  const targetRef = useRef<T>(null);

  return useMemo(() => {
    if (refs.every((ref) => ref == null)) {
      return null;
    }

    refs.forEach((ref) => {
      if (typeof ref === 'function') {
        ref(targetRef.current);
      } else if (isMutableRefObject(ref)) {
        // eslint-disable-next-line no-param-reassign
        ref.current = targetRef.current;
      }
    });

    return targetRef;
  }, [refs]);
}
