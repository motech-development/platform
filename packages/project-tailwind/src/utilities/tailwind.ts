import clsx, { ClassValue } from 'clsx';

/** Alignment options */
export enum Alignment {
  /** Centre */
  CENTRE = 'centre',
  /** Left */
  LEFT = 'left',
  /** Right */
  RIGHT = 'right',
}

/** Alignment type */
export type TAlignment = `${Alignment}`;

/** Sizing options */
export enum Sizing {
  /** Large */
  LG = 'lg',
  /** Medium */
  MD = 'md',
  /** None */
  NONE = 'none',
  /** Small */
  SM = 'sm',
}

/** Sizing type */
export type TSizing = `${Sizing}`;

/** Theming options */
export enum Themes {
  /** Danger */
  DANGER = 'danger',
  /** Primary */
  PRIMARY = 'primary',
  /** Secondary */
  SECONDARY = 'secondary',
  /** Success */
  SUCCESS = 'success',
  /** Warning */
  WARNING = 'warning',
}

/** Theme type */
export type TTheme = `${Themes}`;

/** Class names options */
type TCreateStylesClassNames = {
  /** Universal classes to apply */
  classNames?: ClassValue[];
  /** Sizing specific classes */
  sizing?: {
    [name in Sizing]?: ClassValue[];
  };
  /** Theme specific classes */
  theme?: {
    [name in TTheme]?: ClassValue[];
  };
};

/**
 * Tailwind hook
 *
 * @param theme - The theme to apply
 * @param sizing - Sizing to apply
 *
 * @returns Tailwind helper utilities
 */
export function useTailwind(theme: TTheme, sizing: TSizing) {
  /**
   * Creates css styles to apply
   *
   * @param classNames - Classes to be applied
   *
   * @returns Class names
   */
  function createStyles(classNames: TCreateStylesClassNames) {
    let finalStyles: ClassValue[] = [];

    if (classNames.classNames) {
      finalStyles = [...classNames.classNames];
    }

    if (sizing && classNames.sizing) {
      const styles = classNames.sizing[sizing];

      if (!styles) {
        throw new Error('No classes defined for this sizing.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    if (theme && classNames.theme) {
      const styles = classNames.theme[theme];

      if (!styles) {
        throw new Error('No classes defined for this theme.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    return clsx(finalStyles);
  }

  /**
   * Creates text alignment classes
   *
   * @param alignment - Text alignment
   *
   * @returns Class names
   */
  function createTextAlignmentStyles(alignment: TAlignment) {
    switch (alignment) {
      case Alignment.CENTRE:
        return 'text-center';
      case Alignment.RIGHT:
        return 'text-right';
      case Alignment.LEFT:
      default:
        return 'text-left';
    }
  }

  return {
    createStyles,
    createTextAlignmentStyles,
  };
}
