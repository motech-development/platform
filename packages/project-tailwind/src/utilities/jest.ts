/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';
import { ReactElement } from 'react';
import { Sizing, Themes } from './tailwind';

export const themes = [
  {
    theme: Themes.DANGER,
  },
  {
    theme: Themes.PRIMARY,
  },
  {
    theme: Themes.SECONDARY,
  },
  {
    theme: Themes.SUCCESS,
  },
  {
    theme: Themes.WARNING,
  },
];

export const sizing = [
  {
    size: Sizing.LG,
  },
  {
    size: Sizing.MD,
  },
  {
    size: Sizing.NONE,
  },
  {
    size: Sizing.SM,
  },
];

type TSetup = RenderResult & {
  user: UserEvent;
};

/**
 * Combines Testing Library render function with user events setup
 *
 * @param ui - React element to test
 *
 * @returns Rendered component and test user
 */
export function setup(ui: ReactElement): TSetup {
  return {
    user: userEvent.setup({
      delay: null,
    }),
    ...render(ui),
  };
}
