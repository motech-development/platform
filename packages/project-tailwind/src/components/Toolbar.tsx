import { Slot } from '@radix-ui/react-slot';
import { ComponentPropsWithoutRef, forwardRef } from 'react';
import { Themes, TTheme, useTailwind } from '../utilities/tailwind';
import { ISlot, Slottable } from './Slot';

/** Default component element */
const DEFAULT_ELEMENT = 'div';

/** Toolbar component props */
export interface IToolbarProps extends ISlot, ComponentPropsWithoutRef<'div'> {
  /** Apply border */
  border?: boolean;

  /** Fix position */
  fixed?: boolean;

  /** Component theme */
  theme?: TTheme;
}

/**
 * Flexible application toolbar
 *
 * @param props - Component props
 *
 * @returns Toolbar componenet
 */
export const Toolbar = forwardRef<HTMLDivElement, IToolbarProps>(
  (
    {
      asChild,
      border = false,
      children,
      fixed = false,
      theme = Themes.SECONDARY,
      ...rest
    },
    ref,
  ) => {
    const { createStyles } = useTailwind(theme);

    const Comp = asChild ? Slot : DEFAULT_ELEMENT;

    const outerStyles = createStyles({
      classNames: [
        {
          'fixed top-0 left-0 w-full': fixed,
          relative: !fixed,
        },
      ],
      theme: {
        danger: ['bg-red-900 text-red-50'],
        primary: ['bg-blue-900 text-blue-50'],
        secondary: ['bg-gray-900 text-gray-50'],
        success: ['bg-green-900 text-green-50'],
        warning: ['bg-yellow-900 text-yellow-50'],
      },
    });

    const innerStyles = createStyles({
      classNames: [
        {
          'border-b': border,
        },
      ],
      theme: {
        danger: ['border-red-800'],
        primary: ['border-blue-800'],
        secondary: ['border-gray-800'],
        success: ['border-green-800'],
        warning: ['border-yellow-800'],
      },
    });

    return (
      <Comp className={outerStyles} ref={ref} {...rest}>
        <Slottable asChild={asChild} child={children}>
          {(child) => (
            <div className="mx-auto max-w-7xl select-none sm:px-6 lg:px-8">
              <div className={innerStyles}>
                <div className="flex h-16 items-center justify-between px-4 sm:px-0">
                  {child}
                </div>
              </div>
            </div>
          )}
        </Slottable>
      </Comp>
    );
  },
);
