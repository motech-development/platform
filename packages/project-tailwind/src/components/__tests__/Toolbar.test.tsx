import { setup, themes } from '../../utilities/jest';
import { Toolbar } from '../Toolbar';

const elements = [
  {
    asChild: false,
    element: 'div',
    render: <p>Hello, world</p>,
  },
  {
    asChild: true,
    element: 'header',
    render: (
      <header>
        <p>Hello, world</p>
      </header>
    ),
  },
];

describe('Toolbar', () => {
  describe.each(elements)('as element "$element"', ({ asChild, render }) => {
    describe.each(themes)('when theme is "$theme"', ({ theme }) => {
      it('should correctly render componenet', () => {
        const { asFragment } = setup(
          <Toolbar asChild={asChild} theme={theme}>
            {render}
          </Toolbar>,
        );

        expect(asFragment()).toMatchSnapshot();
      });

      it('should correctly render componenet when border is set', () => {
        const { asFragment } = setup(
          <Toolbar asChild={asChild} border theme={theme}>
            {render}
          </Toolbar>,
        );

        expect(asFragment()).toMatchSnapshot();
      });

      it('should correctly render componenet when position is fixed', () => {
        const { asFragment } = setup(
          <Toolbar asChild={asChild} fixed theme={theme}>
            {render}
          </Toolbar>,
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
