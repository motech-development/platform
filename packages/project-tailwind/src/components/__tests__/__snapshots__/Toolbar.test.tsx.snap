// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar when theme is "danger" should correctly render componenet 1`] = `
<DocumentFragment>
  <div
    class="relative bg-red-900 text-red-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-red-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "danger" should correctly render componenet when border is set 1`] = `
<DocumentFragment>
  <div
    class="relative bg-red-900 text-red-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-red-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "danger" should correctly render componenet when position is fixed 1`] = `
<DocumentFragment>
  <div
    class="fixed top-0 left-0 w-full bg-red-900 text-red-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-red-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "primary" should correctly render componenet 1`] = `
<DocumentFragment>
  <div
    class="relative bg-blue-900 text-blue-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-blue-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "primary" should correctly render componenet when border is set 1`] = `
<DocumentFragment>
  <div
    class="relative bg-blue-900 text-blue-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-blue-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "primary" should correctly render componenet when position is fixed 1`] = `
<DocumentFragment>
  <div
    class="fixed top-0 left-0 w-full bg-blue-900 text-blue-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-blue-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "secondary" should correctly render componenet 1`] = `
<DocumentFragment>
  <div
    class="relative bg-gray-900 text-gray-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-gray-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "secondary" should correctly render componenet when border is set 1`] = `
<DocumentFragment>
  <div
    class="relative bg-gray-900 text-gray-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-gray-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "secondary" should correctly render componenet when position is fixed 1`] = `
<DocumentFragment>
  <div
    class="fixed top-0 left-0 w-full bg-gray-900 text-gray-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-gray-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "success" should correctly render componenet 1`] = `
<DocumentFragment>
  <div
    class="relative bg-green-900 text-green-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-green-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "success" should correctly render componenet when border is set 1`] = `
<DocumentFragment>
  <div
    class="relative bg-green-900 text-green-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-green-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "success" should correctly render componenet when position is fixed 1`] = `
<DocumentFragment>
  <div
    class="fixed top-0 left-0 w-full bg-green-900 text-green-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-green-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "warning" should correctly render componenet 1`] = `
<DocumentFragment>
  <div
    class="relative bg-yellow-900 text-yellow-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-yellow-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "warning" should correctly render componenet when border is set 1`] = `
<DocumentFragment>
  <div
    class="relative bg-yellow-900 text-yellow-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-yellow-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar when theme is "warning" should correctly render componenet when position is fixed 1`] = `
<DocumentFragment>
  <div
    class="fixed top-0 left-0 w-full bg-yellow-900 text-yellow-50"
  >
    <div
      class="mx-auto max-w-7xl select-none sm:px-6 lg:px-8"
    >
      <div
        class="border-yellow-800"
      >
        <div
          class="flex h-16 items-center justify-between px-4 sm:px-0"
        >
          <p>
            Hello, world
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
